Adds empty hand interactions and makes various changes/fixes

Itsnotjames opened the pull request:
Fixes knifes and scalpels not working on animals/players.
Adds CPR.
Adds disarms.
Adds hugs.
Adds punches.
Adds the method InventoryManager.DropItemInSlot(). This drops the item in a specific slot (needed for disarm).
Changes PostToChatMessage.SendItemAttackMessage to SendAttackMessage. This makes it more generic so it can now be used for punches and other attacks in the future.

### Purpose
Fixes #1845
Fixes #1941

### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [X]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [X]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

Itsnotjames committed:
Adds empty hand interactions and makes various changes/fixes

Fixes knifes and scalpels not working on animals/players.
Adds CPR.
Adds disarms.
Adds hugs.
Adds punches.
Adds the method InventoryManager.DropItemInSlot(). This drops the item in a specific slot (needed for disarm).
Changes PostToChatMessage.SendItemAttackMessage to SendAttackMessage. This makes it more generic so it can now be used for punches and other attacks in the future.

DooblyNoobly committed:
Merge branch 'develop' into emptyhandinteractions

fomalsd commented on UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs:
```diff
@@ -158,12 +160,79 @@ public void CmdUnloadWeapon(GameObject gunObject)
+			}
+
+			// Make a random punch hit sound.
+			switch (rng.Next(1, 4))
```
```suggestion
SoundManager.PlayNetworkedAtPos("Punch"+rng.Next(1, 4), victimRegisterTile.WorldPosition);
```
;)

fomalsd commented on UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs:
```diff
@@ -137,6 +138,7 @@ public void CmdUnloadWeapon(GameObject gunObject)
 				playerMove.allowInput = false;
 				attackTarget.Harvest();
 				SoundManager.PlayNetworkedAtPos( "BladeSlice", transform.position );
+				SoundManager.PlayNetworkedAtPos("BladeSlice", transform.position);
```
why do it twice?

Itsnotjames replied:
Thats a mistake

Itsnotjames committed:
Update UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs

Co-Authored-By: Foma <fomalsd@users.noreply.github.com>

Itsnotjames committed:
Remove previous punch hit sound code

Itsnotjames committed:
Remove repeated line

fomalsd commented on UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs:
```diff
@@ -158,12 +160,79 @@ public void CmdUnloadWeapon(GameObject gunObject)
+			ChatRelay.Instance.AddToChatLogServer(new ChatEvent
+			{
+				channels = ChatChannel.Local,
+				message = $"{attackerName} has attempted to punch {victimName}!"
```
apparently attackerName is empty
(and is it needed at all?)
![img](https://user-images.githubusercontent.com/10403536/60588173-c30b0000-9d85-11e9-96f0-5cd13d24a06f.png)

Itsnotjames replied:
![0mqzssR](https://user-images.githubusercontent.com/28093344/60644714-08652580-9e8b-11e9-98b8-ad739a7c149a.png)
Fixed in next commit

fomalsd commented on UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs:
```diff
@@ -204,7 +204,7 @@ public void CmdUnloadWeapon(GameObject gunObject)
 
 			// Make a random punch hit sound.
-			switch (rng.Next(1, 4))
+SoundManager.PlayNetworkedAtPos("Punch"+rng.Next(1, 4), victimRegisterTile.WorldPosition);
```
nonono, that wouldn't compile unless you remove switch case below

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Meleeable.cs:
```diff
@@ -24,18 +24,34 @@ private void Start()
 
 			//special case
-			//We don't melee if we are wielding a gun with ammo and clicking ourselves (we will instead shoot ourselves)
+			//We don't melee if wse are wielding a gun with ammo and clicking ourselves (we will instead shoot ourselves)
```
hehe, introducing typos

Itsnotjames replied:
Fixed

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -869,4 +871,140 @@ public void CmdRequestPaperEdit(GameObject paper, string newMsg)
+	/// Performs a disarm attempt from one player to another.
+	/// </summary>
+	[Command]
+	public void CmdRequestDisarm(GameObject disarmer, GameObject playerToDisarm)
```
apparently you can disarm other people while being dead, lol
also for some reason disarmed players keep their inhand sprites (at least when tested in multiplayer) so I'm not sure if `InventoryManager.DropItemInSlot()` is correct

Itsnotjames replied:
Ah turns out the function that drops an item in a specific slot is PlayerNetworkActions.DropItem(slot).
Fixed this in next commit.

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -869,4 +871,140 @@ public void CmdRequestPaperEdit(GameObject paper, string newMsg)
+	private void DoCPR(GameObject rescuer, GameObject CardiacArrestPlayer)
+	{
+		var CardiacArrestPlayerRegister = CardiacArrestPlayer.GetComponent<RegisterPlayer>();
+		CardiacArrestPlayer.GetComponent<PlayerHealth>().bloodSystem.oxygenDamage -= 7f;
```
Idk what are classic values for CPR, but it's really hard to revive a person if you're the only one performing CPR
if it's supposed to be this way, it's fine, otherwise maybe double this value?

Itsnotjames replied:
It is a bit low, I don't think CPR is very good.
In tg's /mob/living/carbon/human/proc/do_cpr(mob/living/carbon/C) it heals 7 oxy damage until oxy damage is 0.
Here is a picture with the healing code highlighted:
![Nz1lupd](https://user-images.githubusercontent.com/28093344/60640611-0d6ea880-9e7c-11e9-98ed-71b764f2a294.png)
Which goes to:
![zxqNqEk](https://user-images.githubusercontent.com/28093344/60640708-8ec63b00-9e7c-11e9-90bd-7898b1762001.png)

Aranclanos replied:
remember that you can spam cpr on a person and only get the mob clicking cd, you can do several cpr at the same time

fomalsd reviewed (changes requested):
Really good stuff! just needs a couple small bugs fixed, and stricter validation in some places (so you couldn't rob others as a dead person etc)

Itsnotjames committed:
Adds more validation to empty hand interactions, various fixes

Adds stun/down/dead/unconscious/selftarget validation to empty hand interactions.
Fixes typo.
Fixes puncher's name in chat for missed punches.
Fixes dropping items with disarm. They now use the correct function so the sprite dissapears when the item is dropped.

fomalsd reviewed (approved):
Very nice!

DooblyNoobly committed:
Merge branch 'develop' into emptyhandinteractions
